app/support/auto_expire_reservation.rb

```diff
@@ -0,0 +1,37 @@
+class AutoExpireReservation
+  def perform
+    order_details.each do |od|
+      od.transaction do
+        expire_reservation(od)
+      end
+    end
+  end
+
+  private
+
+  def order_details
+    purchased_active_order_details
+  end
+
+  def purchased_active_order_details
+    OrderDetail.purchased_active_reservations
+      .merge(Reservation.relay_in_progress)
+      .where("reservations.reserve_end_at < ?", Time.zone.now - 12.hours)
+      .readonly(false)
+      .all
+  end
+
+  def expire_reservation(od)
+    od.complete!
+
+    # OrderDetail#complete! sets fulfilled_at and price policy,
+    # so we have to reset them.
+    od.fulfilled_at = od.reservation.reserve_end_at
+    od.clear_costs
```

Comment:
Yes, you're right. I'll remove the #clear_costs and also make this method private.